### PR TITLE
feat: ui5-middleware-stringreplacer

### PR DESCRIPTION
**Pull request by petermuessig**
Introduction of a custom string replacer middleware similar to the custom task. Aligned implementations between task and middleware.

### CODE REVIEW

**Comment by petermuessig** [comment]
Follow-up of PR https://github.com/petermuessig/ui5-ecosystem-showcase/pull/273 from @attila-boros

**Comment by petermuessig** [comment]
@attila-boros @TheVivekGowda - please also take a look into this change - it harmonizes the implementation of the custom string replacer task and middleware! 

Thanks!

File: packages/ui5-middleware-stringreplacer/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# UI5 middleware for live string replace in files
+
+Middleware for [ui5-server](https://github.com/SAP/ui5-server), doing a live sring replace on files matched by the includePatterns configuration option.
```
Comment (TheVivekGowda) [inline packages/ui5-middleware-stringreplacer/README.md:3]:
```suggestion
Middleware for [ui5-server](https://github.com/SAP/ui5-server), doing a live string replace on files matched by the includePatterns configuration option.
```

Comment (petermuessig) [reply packages/ui5-middleware-stringreplacer/README.md]: Thanks for the hint! 👍

**Comment by vobu** [comment]
will take a look until eob today

File: packages/ui5-middleware-stringreplacer/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# UI5 middleware for live string replace in files
+
+Middleware for [ui5-server](https://github.com/SAP/ui5-server), doing a live string replace on files matched by the includePatterns configuration option.
```
Comment (vobu) [inline packages/ui5-middleware-stringreplacer/README.md:3]: includePatterns -> `files` array configuration option

File: packages/ui5-app/ui5.yaml [hunk 3]
```diff
@@ -23,7 +23,9 @@ builder:
   - name: ui5-task-stringreplacer
     afterTask: replaceVersion
     configuration:
-      files: ["**/*.js", "**/*.xml"]
+      files:
+        - "**/*.js"
+        - "**/*.xml"
```
Comment (vobu) [inline packages/ui5-app/ui5.yaml:28]: can you please add some replacement configuration as described in the task's/middleware's README?

File: packages/ui5-app/ui5.yaml [hunk 4]
```diff
@@ -68,6 +70,13 @@ server:
       transpileAsync: true
       excludePatterns:
       - "lib/"
+  - name: ui5-middleware-stringreplacer
+    afterMiddleware: compression
+    configuration:
+      debug: true
+      files:
+        - "**/*.js"
+        - "**/*.xml"
```
Comment (vobu) [inline packages/ui5-app/ui5.yaml:79]: can you please add some replacement configuration as described in the task's/middleware's README?

File: packages/ui5-middleware-stringreplacer/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      replace:
+        - placeholder: ${project.artifactId}
+          value: my.sample.app
+        - placeholder: ${project.version}
+          value: 1.0.0-SNAPSHOT
+        - placeholder: '{{app.AppTitle}}'
+          value: My Sample App
+```
+
+3. Maintain all string placeholders and values in `.env` file
```
Comment (vobu) [inline packages/ui5-middleware-stringreplacer/README.md:62]: is this in addition to the configuration listed in 2. ? or only valid for the `task`, not the middleware?

File: packages/ui5-task-stringreplacer/README.md [hunk 6]
```diff
@@ -49,13 +52,25 @@ builder:
...
+      replace:
+        - placeholder: ${project.artifactId}
+          value: my.sample.app
+        - placeholder: ${project.version}
+          value: 1.0.0-SNAPSHOT
+        - placeholder: '{{app.AppTitle}}'
+          value: My Sample App
 ```
+
 3. Maintain all string placeholders and values in `.env` file
```
Comment (vobu) [inline packages/ui5-task-stringreplacer/README.md:67]: same comment as for the middleware - not sure how this fits in addition to the config in 2.

**Review by vobu — CHANGES_REQUESTED** [review]
minor stuff only ;)

**Comment by attila-boros** [comment]
Thanks a lot @petermuessig , I'm very happy to see this. :shipit:

File: packages/ui5-middleware-stringreplacer/README.md (hunk 2 above)
Comment (petermuessig) [reply packages/ui5-middleware-stringreplacer/README.md:3]: Ack

File: packages/ui5-app/ui5.yaml (hunk 3 above)
Comment (petermuessig) [reply packages/ui5-app/ui5.yaml:28]: Ack

File: packages/ui5-app/ui5.yaml (hunk 4 above)
Comment (petermuessig) [reply packages/ui5-app/ui5.yaml:79]: Ack

File: packages/ui5-middleware-stringreplacer/README.md (hunk 5 above)
Comment (petermuessig) [reply packages/ui5-middleware-stringreplacer/README.md:62]: Both configuration options are possible at the same time and will be merged. You can either use the `.env` or the `ui5.yaml` to provide placeholder values.

File: packages/ui5-task-stringreplacer/README.md (hunk 6 above)
Comment (petermuessig) [reply packages/ui5-task-stringreplacer/README.md:67]: Both configuration options are possible at the same time and will be merged. You can either use the `.env` or the `ui5.yaml` to provide placeholder values.

**Comment by petermuessig** [comment]
@vobu - sorry need to again create a force push since I needed to rebase the change on the current master - but the feature looks good now - added your suggestions and thanks to your feedback I did a deeper test and improved the change further... I could already fix some bugs 😉

**Comment by vobu** [comment]
sorry, "Maloche" left and right

**Pull request merged**